File: src/connected_components/WeightedQuickUnion.java
public class WeightedQuickUnion {
    private int[] id, sz;
    
    public WeightedQuickUnion(int n) {
        id = new int[n];
        sz = new int[n];
        for(int i = 0; i < n; i++) {
            id[i] = i;
            sz[i] = 1;
        }
    }

    private int rootOf(int i) {
        while (i != id[i]) {
            id[i] = id[id[i]];  // path compression heuristic
            i = id[i];
        }
        return i;
    }

    public boolean isConnected(int i, int j) {
        return rootOf(i) == rootOf(j);
    }

    public void union(int i, int j) {
        // get root of i, root of j
        // smaller tree goes under bigger tree
        int iRoot = rootOf(i);
        int jRoot = rootOf(j);

        // root of i's tree is smaller
        if (sz[iRoot] < sz[jRoot]) {
            id[iRoot] = jRoot;
            sz[jRoot] += sz[iRoot];
        } else { 
            id[jRoot] = iRoot;
            sz[iRoot] += sz[jRoot];
        }
    }
}

File: src/connected_components/WeightedQuickUnion.rb
class WeightedQuickUnion
  def initialize(num)
    @id = Array.new(num)
    @sz = Array.new(num, 1)
    @id.each_with_index do |i, index|
      @id[index] = index
    end
  end
  
  def rootOf i
    while i != id[i] do
      id[i] = id[id[i]]
      i = id[i]
    end
    i
  end

  def connected?(i, j)
    rootOf i == rootOf j
  end

  def union(i, j)
    iRoot = rootOf i
    jRoot = rootOf j
    if @sz[iRoot] <= @sz[jRoot]
      @id[iRoot] = jRoot
      @sz[jRoot] += @sz[iRoot]
    else
      @id[jRoot] = iRoot
      @sz[iRoot] += @sz[jRoot]
    end
  end
end